### ISSUE TITLE
CI/CD

### ISSUE THREAD

**Issue by StoneyJackson** [ourPLCC/plcc#44]
The `main` branch of the repository contains all the updates since the last release (so the "optional colon", the "add missing new line to some Java files", "print to stdout", etc.). I just need to cut a new release. The problem is, I'm struggling to do so with the release system I created a couple (three?) years ago. I've learned a lot since then, and I think I can do better. So I'm writing this issue to remind myself to rebuild the CI/CD workflows. I'm going to base them on other projects I've done this for more recently (namely https://github.com/HFOSSedu/KitClient).

**Comment by StoneyJackson** [comment]
TLDR: Eliminate VERSION and CHANGLEOG.md relying on GitHub's release entry mechanism instead, and adopt a new development cycle that creates a release automatically each time a pull-request is merged into main.

GOAL: When a PR is merged into main a release will automatically be cut.

RATIONALE: By automating releases, they become part of the normal development cycle and not an extra special step (which folks forget how to do, or are not sure if they work anymore, do not perform consistently, etc.). It ensures the latest and greatest features are available to folks immediately after they have been implemented and tested.

CHALLENGES: Automation depends on triggering actions based on events in the repository: typically whenever a commit is pushed to a particular branch. Our current release process requires two commits to be created as a byproduct of releasing. But that could lead to an infinite loop: "when a commit is pushed to main, create a release by pushing two new commits to main".  There are ways around this, but I would rather simplify our release process and therefore the new automated release process as well.

So let's try to eliminate the two commits created as part of current release process. The first commit is the "release commit" and the second is the "development commit" and is the first development commit in the next release. The "release commit" updates the version number in VERSION and CHANGELOG.md and also contains a summary of changes since the last release. The "development commit" again updates VERSION and CHANGELOG.md with a development version. This ensures that subsequent change that have not been released fall under this development version number and not an official release number.

Under our new automated release mechanism, the "development commit" is no longer needed. Why? Any branch merged into main constitutes a release. So all of the commits that are reachable from main are commits in a release. That's different that how we operate now in which there may be commits in main that have not been release. No more. If it's in main, it's released.

Eliminating the first commit, the "release commit", is a little more challenging. Basically, we need to stop recording the release number in our source code while under version control. So we need to somehow eliminate VERSION and CHANGELOG.md.

We can eliminate VERSION as follows. If we assume that PLCC is always deployed using git, then the current version number can be accessed through a git command. So we update `plcc --version` to do that. If we want to support zip file distributions and such, then we need to update the packaging of these files to include a VERSION number. GitHub (and GitLab) provide mechanisms for publishing a "release entry" which can include additional artifacts like zip files and such. And these "release entries" can be populated automatically as part of the release process. It would take some work to create.

CHANGELOG.md can be eliminated in a similar way. We can use GitHub's "release entry" to summarize the changes since the last release. This too can be automated if we always write a good summary of the change that we are about to merge into main (which creates the release). If we forget to do this, the "release entry" can be manually updated.

The drawback of using GitHub's "release entry" is that it is not portable to other services. So if we moved to another service (like GitLab or Bitbucket) we either hope that the other service provides a way to import this information. If not, we may be able to use GitHub's API to generate a CHANGELOG.md based on the release entries (a tool may already exist to do this).

SUMMARY: So this is my plan. Eliminate VERSION and CHANGLEOG.md relying on GitHub's release entry mechanism to simplify our release process and allow it to be automated.

**Comment by fosler** [comment]
What Stoney says to Tim

"Eliminate VERSION and CHANGLEOG.md relying on GitHub's release entry
mechanism instead, and adopt a new development cycle that creates a
release automatically each time a pull-request is merged into main...

We can eliminate VERSION as follows. If we assume that PLCC is always
deployed using git, then the current version number can be accessed
through a git command. So we update plcc --version to do that. If we
want to support zip file distributions and such, then we need to update
the packaging of these files to include a VERSION number. GitHub (and
GitLab) provide mechanisms for publishing a "release entry" which can
include additional artifacts like zip files and such. And these
"release entries" can be populated automatically as part of the release
process. It would take some work to create...

SUMMARY: So this is my plan. Eliminate VERSION and CHANGLEOG.md relying
on GitHub's release entry mechanism to simplify our release process and
allow it to be automated."

What Tim hears:

"blah blah blah PLCC blah blah blah blah"

**Comment by StoneyJackson** [comment]
@fosler I'll take that as approval :)

**Comment by StoneyJackson** [comment]
Oh... and just so you know... These are issues you are responding to. I use issues to "think out loud". It's part of my design and planning process. Feel free to adjust your notifications on GitHub if you'd rather not be emailed every time I make a comment or an issue :).
